File: state/php/State.php
<?php

// State interface
interface DocumentState {
    public function publish(Document $doc): string;
    public function review(Document $doc): string;
}

// Context
class Document {
    private $state;
    private $name;

    public function __construct(string $name) {
        $this->name = $name;
        $this->state = new DraftState();
    }

    public function setState(DocumentState $state): void {
        $this->state = $state;
    }

    public function publish(): string {
        return $this->state->publish($this);
    }

    public function review(): string {
        return $this->state->review($this);
    }
}

// Concrete States
class DraftState implements DocumentState {
    public function publish(Document $doc): string {
        return "Cannot publish from Draft. Please review first.";
    }

    public function review(Document $doc): string {
        $doc->setState(new ReviewState());
        return "Document moved to Review state";
    }
}

class ReviewState implements DocumentState {
    public function publish(Document $doc): string {
        $doc->setState(new PublishedState());
        return "Document published successfully";
    }

    public function review(Document $doc): string {
        return "Document is already in Review state";
    }
}

class PublishedState implements DocumentState {
    public function publish(Document $doc): string {
        return "Document is already published";
    }

    public function review(Document $doc): string {
        return "Cannot review a published document";
    }
}

// Usage
$doc = new Document("Design Patterns Guide");

echo "Initial state: Draft\n";
echo $doc->publish() . "\n";
echo $doc->review() . "\n";
echo $doc->publish() . "\n";

File: patterns/state/python/State.py
from __future__ import annotations
from abc import ABC, abstractmethod

# State interface
class DocumentState(ABC):
    @abstractmethod
    def publish(self, doc: "Document") -> str:
        pass

    @abstractmethod
    def review(self, doc: "Document") -> str:
        pass

# Context
class Document:
    def __init__(self, name: str):
        self.name = name
        self._state: DocumentState = DraftState()

    def set_state(self, state: DocumentState):
        self._state = state

    def publish(self) -> str:
        return self._state.publish(self)

    def review(self) -> str:
        return self._state.review(self)

# Concrete States
class DraftState(DocumentState):
    def publish(self, doc: Document) -> str:
        return "Cannot publish from Draft. Please review first."

    def review(self, doc: Document) -> str:
        doc.set_state(ReviewState())
        return "Document moved to Review state"

class ReviewState(DocumentState):
    def publish(self, doc: Document) -> str:
        doc.set_state(PublishedState())
        return "Document published successfully"

    def review(self, doc: Document) -> str:
        return "Document is already in Review state"

class PublishedState(DocumentState):
    def publish(self, doc: Document) -> str:
        return "Document is already published"

    def review(self, doc: Document) -> str:
        return "Cannot review a published document"


if __name__ == "__main__":
    doc = Document("Design Patterns Guide")

    print("Initial state: Draft")
    print(doc.publish())
    print(doc.review())
    print(doc.publish())
